clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java

```diff
@@ -0,0 +1,238 @@
+     * inflight.
+     *
+     * @param topic to be requested. If empty, return the metadata for all topics.
+     * @return the future of the metadata request.
+     */
+    public CompletableFuture<Map<String, List<PartitionInfo>>> requestTopicMetadata(final Optional<String> topic) {
+        if (inflightRequests.containsKey(topic)) {
+            return inflightRequests.get(topic).future;
+        }
+
+        TopicMetadataRequestState newRequest = new TopicMetadataRequestState(
+                logContext,
+                topic,
+                retryBackoffMs,
+                retryBackoffMaxMs);
+        inflightRequests.put(topic, newRequest);
+        return newRequest.future;
+    }
+
+    // Visible for testing
+    List<TopicMetadataRequestState> inflightRequests() {
+        return new ArrayList<>(inflightRequests.values());
+    }
+
+    class TopicMetadataRequestState extends RequestState {
+        private final Optional<String> topic;
+        CompletableFuture<Map<String, List<PartitionInfo>>> future;
+
+        public TopicMetadataRequestState(final LogContext logContext,
+                                         final Optional<String> topic,
+                                         final long retryBackoffMs,
+                                         final long retryBackoffMaxMs) {
+            super(logContext, TopicMetadataRequestState.class.getSimpleName(), retryBackoffMs,
+                retryBackoffMaxMs);
+            this.future = new CompletableFuture<>();
+            this.topic = topic;
+        }
+
+        /**
+         * prepare the metadata request and return an
+         * {@link org.apache.kafka.clients.consumer.internals.NetworkClientDelegate.UnsentRequest} if needed.
+         */
+        private Optional<NetworkClientDelegate.UnsentRequest> send(final long currentTimeMs) {
+            if (!this.canSendRequest(currentTimeMs)) {
+                return Optional.empty();
+            }
+            this.onSendAttempt(currentTimeMs);
+
+            final MetadataRequest.Builder request;
+            request = topic.map(t -> new MetadataRequest.Builder(Collections.singletonList(t), allowAutoTopicCreation))
+                    .orElseGet(MetadataRequest.Builder::allTopics);
+
+            final NetworkClientDelegate.UnsentRequest unsent = createUnsentRequest(request);
+            return Optional.of(unsent);
+        }
+
+        private NetworkClientDelegate.UnsentRequest createUnsentRequest(
+                final MetadataRequest.Builder request) {
+            return new NetworkClientDelegate.UnsentRequest(
+                    request,
+                    Optional.empty(),
+                    this::processResponseOrException
+            );
+        }
+
+        private void processResponseOrException(final ClientResponse response,
+                                                final Throwable exception) {
+            long responseTimeMs = System.currentTimeMillis();
```

Comment:
Could we pass in currentTime from the caller?

Reply:
Hi Jun - understood making repeated system calls is bad.  However - we do need to get the time when the response is received, so this is kind of the lazy way to do it.  

I think it is possible to expose the time during the callback/future completion during the network poll (because it is computed at the top of the background thread loop and pass down to network poll.

Reply:
Actually, could we use `ClientResponse.receivedTimeMs`?